program/psiconv/psiconv.man
.TH PSICONV 1 "October 19, 2014" 0.9.9 ""
.SH NAME
psiconv \- Psion 5 and other EPOC devices file converter
.SH SYNOPSIS
.B psiconv
.RI [ options ]
.RI [ file ]
.SH DESCRIPTION
.B psiconv
is used to convert files generated by Psion 5, Psion 5MX and other EPOC devices into more common formats. It can currently convert Word and TextEd files into ASCII, HTML4 or XHTML, and Sketch, MBM and ClipArt into most common picture formats. 

Psiconv works like a filter: by default files are read from stdin and written
to stdout. 
.SH OPTIONS
.TP
.BR \-c , " \-\-configfile \fIfile\fP"
Read an additional configuration file after the default ones. See psiconv.conf(5).
.TP
.BR \-e , " \-\-encoding \fIenc\fP"
Output encoding. Supported encodings are \fBUTF8\fP, \fBUCS2\fP, \fBPsion\fP (the same encoding as the Psion file used) or 7-bit \fBASCII\fP.
.TP
.BR \-h , " \-\-help"
Output a small usage guide as well as the supported output formats on stdout and exit successfully. No other output is generated.
.TP
.BR \-n , " \-\-noise \fIlevel\fP"
The amount of noise generated on stderr. Recognized values are \fB1\fP or \fBF\fP for fatal errors only, \fB2\fP or \fBE\fP for all errors, \fB3\fP or \fBW\fP for errors and warnings, \fB4\fP or \fBP\fP to include progress indicators and \fB5\fP or \fBD\fP for low level debug information.
.TP
.BR \-o , " \-\-outputfile \fIfile\fP"
File to write output to, instead of stdout.
.TP
.BR \-T , " \-\-type \fIfileformat\fP"
The file format to output. By default, Word and TextEd files will be converted to XHTML and Sketch, MBM and Clip Art files to TIFF. The full list of output formats can be found by using the \fB-h\fP option.
.TP
.BR \-V , " \-\-version"
Output the current version of psiconv on stdout and exit successfully. No other output is generated.
.SH AUTHOR
Frodo Looijaard <frodo@frodo.looijaard.name>
.SH "SEE ALSO"
.BR psiconv.conf (5).



NEWS
0.9.9  20141019  Bugfix in HTML4 target. Updated toolchain. 
                 Support for GraphicsMagick. Updated manual page psiconv(1).

0.9.8  20051115  Small bug fixes

0.9.7  20050225  Support ImageMagick 6

0.9.6  20040430  Compilation fixes for Solaris
                 Compilation fixes for ImageMagick unavailability

0.9.5  20040406  Bug fixes
                 Compilation with gcc 2.95.x should work again

0.9.4  20040316  Mostly bug fixes
                 Minimum version required for AbiWord 2.1.1 and later

0.9.3  20040226  Minimum version required for new AbiWord 2.1 plugin 
                 Fixed a few minor bugs
                 Added psiconv_unicode_from_list and psiconv_unicode_strstr
                 Much better error and progress reporting in generators

0.9.2.1 20040223 Quick fix for the library number mistake in 0.9.2

0.9.2  20040223  Many bugfixes, among them some critical ones:
                 * Package now builds without psiconv.conf files installed
                 UTF8 Psion support might now actually work
                 -c argument introduced for psiconv program

0.9.1  20040209  Added the psiconv-config program.
		 Fixed the installation of @includedir@/psiconv/general.h
                 Added the forgotted psiconv_config_free function.
		 Removed some small memory leaks.

0.9.0  20040204  New major library version number breaks source compatibility:
		 * All strings and characters are now stored in UCS-2
                 * New config structure.
                   - Call psiconv_config_default and psiconv_config_read at the
                     start of your program
                   - Almost all functions have a new psiconv_config parameter
                   - All former global variables are now in this structure
                 * Several minor changes in the functions defined in
                   parse_routines.h and generate_routines.h
                 * All generate_* routines have now a lev parameter except
                   generate_file.
                 * psiconv_sketch_section data representation changes
                 * New verbosity level PSICONV_VERB_ERROR
                 Use config files /etc/psiconv.conf and ~/.psiconv.conf
                 Support automake 1.6, 1.7 and 1.8
                 Support autoconf-2.50 and up
                 Support ImageMagick 5.4.x and 5.5.x (API changed once again...)
                 Support Unicode output in psiconv program (Keita Kawabe)
                 Don't use features not present in plain sh (Keita Kawabe)
                 Clean up some automake-related stuff.
		 Format documentation is now installed too.
                 Debian build support
                 Embedded objects in Word are now properly parsed and generated
                 Sketch file generation is added (stand-alone and as object)
                 MBM and Clipart file generation is added
                 Image files other than 2-bits greyscale are parsed and
                   generated (experimental)
                 Errors and warnings sanitized
                 Rewrite of the psiconv program:
                 * Outputs in UTF8, UCS2, ASCII or Psion encodings
		 * XHTML target generates strict XHTML using CSS
                 * HTML4 target generates traditional HTML without CSS
                 * IMAGE target remains the same
                 * LATEX and RTF targets are not supported anymore
0.8.3  20020129  Extended sheet parsing support in library:
                     all non-graph parts are now parsed.
                 Support autoconf-2.52, libtool-1.4.2 (needs patch!), 
                     automake-1.5
                 Support ImageMagick-5.4.2-2 (older may not work)
                 Fix header/footer potential segfault
                 Fix large list-related memory leak and several smaller ones.
0.8.2  20010717  Some sheet parsing support in library, for from complete,
                     but good enough for basic Gnumeric support
                 Added autogen.sh
                 autoconf-2.50, automake-1.4-p4
0.8.1  20010110  New enum screenfont_t
                 New program rewrite in extra
                 Added psiconv_empty_* functions
                 Fixed a few minor generation bugs
0.8.0  20001228  Fixed style inheritance from Normal
                 Fixed warnings in layouts
                 Added generation routines to libpsiconv
                 Upgraded documentation to version 2.6
                 Fixed several minor parse-related issues
0.7.1  20001218  Maintenance release to be used with Abiword.
                 Set proper interline defaults.
0.7.0  20001215  Made libpsiconv C++ compatible.
                 general.h is now installed properly (oops...)
                 Revamped package directory structure and files.
                 Include files contain now the correct paths.
                 Fixed a nasty typo; larger files should now work.
                 It seems psiconv_verbosity was never defined.
                 Added message printing hook for applications to capture.
                 All result codes are now tested, library is much more robust.
                 Some internal data fields have been renamed.
0.6.2  20001021  ImageMagick 5 now works. I hope.
0.6.1  19991204  Added Clipart files. Several pictures in one file are now
                 handled correctly.
0.6.0  19991202  Renamed HTML to HTML3. Images can be handled through the
                 ImageMagick library. Moved the generation routines out of
                 libpsiconv.
0.5.0  19990930  Automake, autoconf and libtool support. No functional changes.
                 One small bug squashed in the parser.
0.4.0  19990812  First release of libpsiconv. It now supports Word and
                 TextEd files. Many internal things have changed, including
                 the names of many functions.
                 Renamed TEXT format to ASCII, and the -t option to -T.
                 Psiconv is now linked against libpsiconv

program/psiconv/psiconv.c
/*
    psiconv.c - Part of psiconv, a PSION 5 file formats converter
    Copyright (c) 1999-2014  Frodo Looijaard <frodo@frodo.looijaard.name>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/* Driver program */

#include "config.h"
#include "compat.h"
#include <getopt.h>
#include <stdio.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef IMAGEMAGICK
#include "magick-aux.h"
#endif

#include <psiconv/data.h>
#include <psiconv/list.h>
#include <psiconv/parse.h>
#include <psiconv/configuration.h>
#include "psiconv.h"
#include "gen.h"

static void print_help(void);
static void print_version(void);
static void strtoupper(char *str);

void print_help(void)
{
  fileformat ff;
  int i,j;

  puts("Syntax: psiconv [OPTIONS..] [FILE]");
  puts("Convert the psion file FILE to other formats");
  puts("If FILE is not specified, use stdin");
  puts("  -c, --configfile=FILE Read extra configuration file after normal ones");
  puts("  -e, --encoding=ENC    Output encoding (default: UTF8)");
  puts("  -h, --help            Display this help and exit");
  puts("  -n, --noise=LEVEL     Select what to print on stderr (overrides psiconv.conf)");
  puts("  -o, --outputfile      Output to file instead of stdout");
  puts("  -T, --type=FILETYPE   Output type (default: XHTML or TIFF");
  puts("  -V, --version         Display the program version and exit");
  puts("");
  puts("The following encodings are currently supported:");
  puts("  UTF8    Variable length Unicode encoding");
  puts("  UCS2    Fixed 16-bit length Unicode encoding");
  puts("  Psion   The encoding your Psion uses (as in psiconv.conf)");
  puts("  ASCII   7-bit ASCII (other symbols are substituted by '?')");
  puts("");
  puts("The following noise levels are currently supported:");
  puts("  1 or F: Fatal errors only");
  puts("  2 or E: Errors");
  puts("  3 or W: Warnings");
  puts("  4 or P: Progress indicators");
  puts("  5 or D: Debug data");
  puts("");
  puts("The following abbreviations are used in the output types list:");
  puts("  C  - processes ClipArt files");
  puts("  c  - processes ClipArt files containing only one image");
  puts("  M  - processes MBM files");
  puts("  m  - processes MBM files containing only one image");
  puts("  S  - processes Sketch files");
  puts("  T  - processes TextEd files");
  puts("  W  - processes Word files");
  puts("");
  puts("The following output types are known:");
  for (i = 0; i < psiconv_list_length(fileformat_list); i ++) {
    ff = psiconv_list_get(fileformat_list,i);
    printf("  %s",ff->name);
    for (j = strlen(ff->name); j < 10; j++)
      putchar(' ');
    printf("[%c%c%c%c%c] ",
	   ff->supported_format & FORMAT_CLIPART_MULTIPLE?'C':
	   ff->supported_format & FORMAT_CLIPART_SINGLE?'c':' ',
	   ff->supported_format & FORMAT_MBM_MULTIPLE?'M':
	   ff->supported_format & FORMAT_MBM_SINGLE?'m':' ',
	   ff->supported_format & FORMAT_SKETCH?'S':' ',
	   ff->supported_format & FORMAT_TEXTED?'T':' ',
	   ff->supported_format & FORMAT_WORD?'W':' ');
    puts(ff->description);
  }
  puts("");
  puts("When using UTF8 with LaTeX type, the resulting LaTeX source should be converted");
  puts(" to a suitable encoding for your LaTeX installation before being typeset");
}

void print_version(void)
{
  printf("Version %s\n",VERSION);
}

void strtoupper(char *str)
{
  int i;
  for (i = 0; i < strlen(str); i ++)
    str[i] = toupper(str[i]);
}

int main(int argc, char *argv[])
{
  struct option long_options[] =
  {
    {"help",no_argument,NULL,'h'},
    {"version",no_argument,NULL,'V'},
    {"configfile",required_argument,NULL,'c'},
    {"noise",required_argument,NULL,'n'},
    {"outputfile",required_argument,NULL,'o'},
    {"type",required_argument,NULL,'T'},
    {"encoding",no_argument,NULL,'e'},
    {0,0,0,0}
  };
  const char* short_options = "hVn:o:e:T:c:";
  int option_index;
  FILE * f;
  struct stat fbuf;

  const char *inputfilename = "";
  const char *outputfilename = "";
  const char *extra_configfile = NULL;
  char *type = NULL;
  encoding encoding_type=ENCODING_UTF8;
  psiconv_list outputlist;
  int verbosity = 0;
  psiconv_config config;

  int c,i,res;
  psiconv_buffer buf;
  psiconv_file file;
  fileformat ff = NULL;

  if (!(fileformat_list = psiconv_list_new(sizeof(struct fileformat_s)))) {
    fputs("Out of memory error",stderr);
    exit(1);
  }

  init_txt();
  init_xhtml();
  init_html4();
  init_html5();
  init_image();

  while(1) {
    c = getopt_long(argc,argv,short_options, long_options, &option_index);
    if (c == -1)
      break;
    switch(c) {
      case 'h': print_help(); exit(0);
      case 'V': print_version(); exit(0);
      case 'n': switch(optarg[0]) {
		  case '1': case 'F':case 'f':
		    verbosity=PSICONV_VERB_FATAL;
		    break;
		  case '2': case 'E':case 'e':
		    verbosity=PSICONV_VERB_ERROR;
		    break;
		  case '3': case 'W':case 'w':
		    verbosity=PSICONV_VERB_WARN;
		    break;
		  case '4': case 'P':case 'p':
		    verbosity=PSICONV_VERB_PROGRESS;
		    break;
		  case '5': case 'D':case 'd':
		    verbosity=PSICONV_VERB_DEBUG;
		    break;
		  default:
		    fputs("Unknown noise level\n",stderr);
		    exit(1);
		}
                break;
      case 'o': outputfilename = strdup(optarg); break;
      case 'T': type = strdup(optarg); break;
      case 'e': if(!strcmp(optarg,"UTF8"))
		  encoding_type = ENCODING_UTF8;
		else if (!strcmp(optarg,"UCS2"))
		  encoding_type = ENCODING_UCS2;
		else if (!strcmp(optarg,"ASCII"))
		  encoding_type = ENCODING_ASCII;
		else if (!strcmp(optarg,"Psion"))
		  encoding_type = ENCODING_PSION;
		else {
		  fputs("Unknown encoding type "
		        "(try '-h' for more information\n",stderr);
		  exit(1);
		}
		break;
      case 'c': extra_configfile = strdup(optarg); break;
      case '?': case ':': fputs("Try `-h' for more information\n",stderr);
                          exit(1);
      default: fprintf(stderr,"Internal error: getopt_long returned character "
                              "code 0%o ?? (contact the author)\n", c);
               exit(1); break;
    }
  }
  if (optind < argc-1) {
    fputs("I can only convert one file!\n"
          "Try `-h' for more information\n",stderr);
    exit(1);
  } else if (optind == argc-1)
    if (!(inputfilename = strdup(argv[optind]))) {
      fputs("Out of memory error",stderr);
      exit(1);
    }

  config = psiconv_config_default();
  psiconv_config_read(extra_configfile,&config);
  if (verbosity)
    config->verbosity = verbosity;

  /* Open inputfile for reading */

  if (strlen(inputfilename) != 0) {
    if(stat(inputfilename,&fbuf) < 0) {
      perror(inputfilename);
      exit(1);
    }
    f = fopen(inputfilename,"r");
    if (! f) {
      perror(inputfilename);
      exit(1);
    }
  } else
    f = stdin;

  if (!(buf = psiconv_buffer_new())) {
    fputs("Out of memory error",stderr);
    exit(1);
  }
  if (psiconv_buffer_fread_all(buf,f)) {
    fprintf(stderr,"Failure reading file");
    exit(1);
  }

  if (strlen(inputfilename) != 0)
    if (fclose(f)) {
      perror(inputfilename);
      exit(1);
    }

  if (psiconv_parse(config,buf,&file) || (file->type == psiconv_unknown_file))
  {
     fprintf(stderr,"Parse error\n");
     exit(1);
  }

  if (!type) {
    switch(file->type) {
      case psiconv_word_file:
      case psiconv_texted_file:
      default:
	type = "XHTML"; break;
      case psiconv_mbm_file:
      case psiconv_clipart_file:
      case psiconv_sketch_file:
	type = "TIFF"; break;
    }
  } else
    strtoupper(type);

  for (i = 0; i < psiconv_list_length(fileformat_list); i ++) {
    ff = psiconv_list_get(fileformat_list,i);
    if (! strcasecmp(type,ff->name)) {
      break;
    }
  }

  if (i == psiconv_list_length(fileformat_list)) {
    fprintf(stderr,"Unknown output type: `%s'\n",type);
    exit(1);
  }

  if (!(outputlist = psiconv_list_new(sizeof(psiconv_u8)))) {
    fputs("Out of memory error\n",stderr);
    exit(1);
  }

  res = ff->output(config,outputlist,file,type,encoding_type);
  if (res) {
    fprintf(stderr,
            "Output format `%s' not permitted for this file type\n",type);
    exit(1);
  }

  psiconv_free_file(file);

  if (strlen(outputfilename) != 0) {
    f = fopen(outputfilename,"w");
    if (! f) {
      perror(inputfilename);
      exit(1);
    }
  } else
    f = stdout;

  psiconv_list_fwrite_all(outputlist,f);

  if (fclose(f)) {
    perror(inputfilename);
    exit(1);
  }

  psiconv_list_free(outputlist);

  exit(0);
}
